File: 7785/7785.py
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#    7785.py                                            :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2023/10/30 22:23:39 by jihoolee          #+#    #+#              #
#    Updated: 2023/10/30 22:30:18 by jihoolee         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

def main() -> None:
    input_cnt: int = int(input())
    present_workers: set = set()

    for _ in range(input_cnt):
        name, info = input().split(' ')
        if (info == 'enter'):
            present_workers.add(name)
        else:
            present_workers.remove(name)

    for name in sorted(present_workers, reverse=True):
        print(name)


if (__name__ == '__main__'):
    main()

File: 7785/7785.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   7785.cpp                                           :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: jihoolee <jihoolee@student.42SEOUL.kr>     +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2023/10/30 21:49:51 by jihoolee          #+#    #+#             */
/*   Updated: 2023/10/30 22:23:32 by jihoolee         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include <set>
#include <string>

void  handle_entry_exit(std::set<std::string>* p_present_workers) {
  std::set<std::string>&  present_workers = *p_present_workers;
  std::string             name;
  std::string             info;

  std::cin >> name >> info;
  if (info == "enter")
    present_workers.insert(name);
  else
    present_workers.erase(name);
}

int main(void) {
  int                   input_cnt;
  std::set<std::string> present_workers;

  std::cin >> input_cnt;
  for (int i = 0; i < input_cnt; ++i)
    handle_entry_exit(&present_workers);
  for (std::set<std::string>::const_reverse_iterator it =
          present_workers.crbegin();
        it != present_workers.crend();
        ++it)
    std::cout << *it << '\n';
  return 0;
}
